src/core/lib/event_engine/windows/win_socket.cc

```diff
@@ -0,0 +1,201 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include <grpc/support/port_platform.h>
+
+#ifdef GPR_WINDOWS
+#include <grpc/support/alloc.h>
+#include <grpc/support/log_windows.h>
+
+#include "src/core/lib/event_engine/executor/executor.h"
+#include "src/core/lib/event_engine/trace.h"
+#include "src/core/lib/event_engine/windows/win_socket.h"
+#include "src/core/lib/gprpp/sync.h"
+#include "src/core/lib/iomgr/error.h"
+
+#if defined(__MSYS__) && defined(GPR_ARCH_64)
+/* Nasty workaround for nasty bug when using the 64 bits msys compiler
+   in conjunction with Microsoft Windows headers. */
+#define GRPC_FIONBIO _IOW('f', 126, uint32_t)
+#else
+#define GRPC_FIONBIO FIONBIO
+#endif
+
+namespace grpc_event_engine {
+namespace experimental {
+
+WinSocket::WinSocket(SOCKET socket, Executor* executor) noexcept
+    : socket_(socket),
+      executor_(executor),
+      read_info_(OpState(this)),
+      write_info_(OpState(this)) {}
+
+WinSocket::~WinSocket() { GPR_ASSERT(is_shutdown_); }
+
+SOCKET WinSocket::socket() { return socket_; }
+
+void WinSocket::MaybeShutdown(absl::Status why) {
+  grpc_core::MutexLock lock(&mu_);
+  // if already shutdown, return early. Otherwise, set the shutdown flag.
+  if (is_shutdown_) {
+    if (GRPC_TRACE_FLAG_ENABLED(grpc_event_engine_trace)) {
+      gpr_log(GPR_DEBUG, "WinSocket::%p already shutting down", this);
+    }
+    return;
+  }
+  is_shutdown_ = true;
+  if (GRPC_TRACE_FLAG_ENABLED(grpc_event_engine_trace)) {
+    gpr_log(GPR_DEBUG, "WinSocket::%p shutting down now. Reason: %s", this,
+            why.ToString().c_str());
+  }
+  // Grab the function pointer for DisconnectEx for that specific socket.
+  // It may change depending on the interface.
+  GUID guid = WSAID_DISCONNECTEX;
+  LPFN_DISCONNECTEX DisconnectEx;
+  DWORD ioctl_num_bytes;
+  int status = WSAIoctl(socket_, SIO_GET_EXTENSION_FUNCTION_POINTER, &guid,
+                        sizeof(guid), &DisconnectEx, sizeof(DisconnectEx),
+                        &ioctl_num_bytes, NULL, NULL);
+
+  if (status == 0) {
+    DisconnectEx(socket_, NULL, 0, 0);
+  } else {
+    char* utf8_message = gpr_format_message(WSAGetLastError());
+    gpr_log(GPR_INFO, "Unable to retrieve DisconnectEx pointer : %s",
+            utf8_message);
+    gpr_free(utf8_message);
+  }
+  closesocket(socket_);
+}
+
+void WinSocket::NotifyOnReady(OpState& info, EventEngine::Closure* closure) {
+  grpc_core::MutexLock lock(&mu_);
+  if (IsShutdown()) {
+    info.SetError(WSAESHUTDOWN);
+    executor_->Run(closure);
+    return;
+  };
+  if (info.has_pending_iocp_) {
```

Comment:
nowadays I'm getting into the habit of writing `if (absl::exchange(info.has_pending_iocp_, false)) {` for this pattern to more clearly & concisely label that it's a state edge

Reply:
Done.